### PR TITLE
Resolve #21 refactoring

### PR DESCRIPTION
**Pull request by aoutir**

### CODE REVIEW

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 1]
```diff
@@ -401,7 +401,7 @@ private DataTypeTest memSqlCharTypeTest()
     @Test
     public void testMemSqlCreatedParameterizedCharUnicode()
     {
-        DataTypeTest.create()
+        DataTypeTestToSqlDataTypeTestConverter.create()
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:404]: If you use this, you need to read the output in the command line. Then use that (some changes may be required), to change DataTypeTest to SQLDataTypeTest.

**Review by Atema — CHANGES_REQUESTED** [review]

**Comment by Atema** [comment]
Take a look at #3. `TestPostgreSqlTypeMapping` in the repo can also be a good help.

**Comment by aoutir** [comment]
Yes but they don't use the class `DataTypeTestToSqlDataTypeTestConverter`, I tried something I am trying to get the tests to run I will marked it ready to review if the tests are good

**Comment by Atema** [comment]
If you run the code that's in your first commit, it should write the 
SqlDataTypeTest code to the standard output. That can be used as a basis 
for the rewritten tests.

------ Original Message ------

**Comment by aoutir** [comment]
SqlDataTypeTest code is also present in the message of the `RuntimeException `thrown in the method `execute `of the class `DataTypeTestToSqlDataTypeTestConverter ` that's why I am calling `getMessage ` to get the code and then I am executing it.
But I have just seen that the message starts with Code replacement so I can't use it directly

**Comment by Atema** [comment]
The code will still need some changes that need to be edited manually

------ Original Message ------

**Comment by aoutir** [comment]
You're right, maybe I will see what I need to change clearly when I run the tests

**Comment by aoutir** [comment]
if you still have the output can you post it here ?

**Comment by Atema** [comment]
```
< FAILURE!
java.lang.RuntimeException:
Code replacement:
SqlDataTypeTest.create()
 .addRoundTrip("char(1) CHARACTER SET utf8", "'攻'", createCharType(1), "'攻'")
 .addRoundTrip("char(5) CHARACTER SET utf8", "'攻殻'", createCharType(5), "'攻殻'") // TODO non-identity toTrinoQueryResult function was used: 攻殻 vs 攻殻
 .addRoundTrip("char(5) CHARACTER SET utf8", "'攻殻機動隊'", createCharType(5), "'攻殻機動隊'")
 .execute( ... ) // TODO

        at io.trino.testing.datatype.DataTypeTestToSqlDataTypeTestConverter.execute(DataTypeTestToSqlDataTypeTestConverter.java:115)
        at io.trino.plugin.memsql.TestMemSqlTypeMapping.testMemSqlCreatedParameterizedCharUnicode(TestMemSqlTypeMapping.java:409)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)

[ERROR] testTrinoCreatedParameterizedVarchar(io.trino.plugin.memsql.TestMemSqlTypeMapping)  Time elapsed: 0.001 s  <<< FAILURE!
java.lang.RuntimeException:
Code replacement:
SqlDataTypeTest.create()
 .addRoundTrip("varchar(10)", "'text_a'", createVarcharType(10), "'text_a'")
 .addRoundTrip("varchar(255)", "'text_b'", createVarcharType(255), "'text_b'")
 .addRoundTrip("varchar(256)", "'text_c'", createVarcharType(256), "'text_c'")
 .addRoundTrip("varchar(21844)", "'text_memsql_max'", createVarcharType(21844), "'text_memsql_max'")
 .addRoundTrip("varchar(21845)", "'text_memsql_larger_than_max'", createVarcharType(65535), "'text_memsql_larger_than_max'")
 .addRoundTrip("varchar(65535)", "'text_d'", createVarcharType(65535), "'text_d'")
 .addRoundTrip("varchar(65536)", "'text_e'", createVarcharType(16777215), "'text_e'")
 .addRoundTrip("varchar(16777215)", "'text_f'", createVarcharType(16777215), "'text_f'")
 .addRoundTrip("varchar(16777216)", "'text_g'", createUnboundedVarcharType(), "'text_g'")
 .addRoundTrip("varchar(2147483646)", "'text_h'", createUnboundedVarcharType(), "'text_h'")
 .addRoundTrip("varchar", "'unbounded'", createUnboundedVarcharType(), "'unbounded'")
 .execute( ... ) // TODO
```

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 2]
```diff
@@ -401,30 +403,32 @@ private DataTypeTest memSqlCharTypeTest()
...
+        .addRoundTrip("varchar(256)", "'text_c'", createVarcharType(256), "'text_c'")
+        .addRoundTrip("varchar(21844)", "'text_memsql_max'", createVarcharType(21844), "'text_memsql_max'")
+        .addRoundTrip("varchar(21845)", "'text_memsql_larger_than_max'", createVarcharType(65535), "'text_memsql_larger_than_max'")
+        .addRoundTrip("varchar(65535)", "'text_d'", createVarcharType(65535), "'text_d'")
+        .addRoundTrip("varchar(65536)", "'text_e'", createVarcharType(16777215), "'text_e'")
+        .addRoundTrip("varchar(16777215)", "'text_f'", createVarcharType(16777215), "'text_f'")
+        .addRoundTrip("varchar(16777216)", "'text_g'", createUnboundedVarcharType(), "'text_g'")
+        .addRoundTrip("varchar(2147483646)", "'text_h'", createUnboundedVarcharType(), "'text_h'")
+        .addRoundTrip("varchar", "'unbounded'", createUnboundedVarcharType(), "'unbounded'")
                 .execute(getQueryRunner(), trinoCreateAsSelect("trino_test_parameterized_varchar"));
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:429]: Indentation should be 16.

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 3]
```diff
@@ -79,6 +79,8 @@
 import static java.math.RoundingMode.HALF_UP;
 import static java.math.RoundingMode.UNNECESSARY;
 import static java.util.Arrays.asList;
+import java.io.*;
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:82]: Remove this

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 4]
```diff
@@ -79,6 +79,8 @@
 import static java.math.RoundingMode.HALF_UP;
 import static java.math.RoundingMode.UNNECESSARY;
 import static java.util.Arrays.asList;
+import java.io.*; 
+
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:83]: Imports missing are createCharType and SqlDataTypeTest.

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 5]
```diff
@@ -401,30 +403,32 @@ private DataTypeTest memSqlCharTypeTest()
     @Test
     public void testMemSqlCreatedParameterizedCharUnicode()
     {
-        DataTypeTest.create()
-                .addRoundTrip(charDataType(1, CHARACTER_SET_UTF8), "\u653b")
-                .addRoundTrip(charDataType(5, CHARACTER_SET_UTF8), "\u653b\u6bbb")
-                .addRoundTrip(charDataType(5, CHARACTER_SET_UTF8), "\u653b\u6bbb\u6a5f\u52d5\u968a")
+        SqlDataTypeTest.create()
+                .addRoundTrip("char(1) CHARACTER SET utf8", "'攻'", createCharType(1), "'攻'")
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:407]: You need to cast here as well (just like next line)

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 6]
```diff
@@ -401,30 +403,32 @@ private DataTypeTest memSqlCharTypeTest()
...
+                .addRoundTrip("char(5) CHARACTER SET utf8", "'攻殻'", createCharType(5), "CAST('攻殻' AS char(5))") 
+                .addRoundTrip("char(5) CHARACTER SET utf8", "'攻殻機動隊'", createCharType(5), "'攻殻機動隊'")
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:409]: Same

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java (hunk 2 above)
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:429]: You need to cast the right-hand side, like in the other function.

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java (hunk 6 above)
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:409]: `CHARACTER SET utf8` may not be nmecessary (not sure)

**Review by Atema — CHANGES_REQUESTED** [review]

**Comment by aoutir** [comment]
the tests are taking so long, they are working except for one I correct it I tried to run them again but they never end.
I guess it's working now it was just about a little detail but you never know run the tests again.

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 7]
```diff
@@ -401,29 +402,28 @@ private DataTypeTest memSqlCharTypeTest()
...
+                .addRoundTrip("varchar(" + MEMSQL_VARCHAR_MAX_LENGTH + ")", "'text_memsql_max'", createVarcharType(MEMSQL_VARCHAR_MAX_LENGTH), "CAST('text_memsql_max' AS varchar(" + MEMSQL_VARCHAR_MAX_LENGTH + "))")
+                .addRoundTrip("varchar(" + (MEMSQL_VARCHAR_MAX_LENGTH + 1) + ")", "'text_memsql_larger_than_max'", createVarcharType(65535), "CAST('text_memsql_larger_than_max' AS varchar(65535))")
+                .addRoundTrip("varchar(65535)", "'text_d'", createVarcharType(65535), "CAST('text_d' AS varchar(65535))")
+                .addRoundTrip("varchar(65536)", "'text_e'", createVarcharType(16777215), "CAST('text_e' AS varchar(65536))")
+                .addRoundTrip("varchar(16777215)", "'text_f'", createVarcharType(16777215), "CAST('text_f' AS varchar(16777215))")
+                .addRoundTrip("varchar(16777216)", "'text_g'", createUnboundedVarcharType(), "CAST('text_g' AS varchar(16777216))")
+                .addRoundTrip("varchar(2147483646)", "'text_h'", createUnboundedVarcharType(), "CAST('text_h' AS varchar(2147483646))")
+                .addRoundTrip("varchar", "'unbounded'", createUnboundedVarcharType(), "CAST('unbounded' AS varchar)")
```
Comment (Atema) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:426]: You're almost done :). The type of the expected literal (4th argument) needs to be the same as the expected type (3rd argument). For example, in line 425 should have `CAST('text_h' AS varchar)`

**Review by Atema — CHANGES_REQUESTED** [review]

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java (hunk 7 above)
Comment (aoutir) [reply plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:426]: I need to change this only for line 424 and 425 right, the rest is good ?

Comment (Atema) [reply plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java:426]: 422 as well

**Comment by aoutir** [comment]
All the tests are working now

**Comment by Atema** [comment]
Great!

**Pull request merged**